Tests/AsyncHTTPClientTests/HTTPClientTests.swift

```diff
@@ -2051,4 +2053,45 @@ class HTTPClientTests: XCTestCase {
 
         XCTAssertNoThrow(try future.wait())
     }
+
+    func testContentLengthTooLongFails() throws {
+        let url = self.defaultHTTPBinURLPrefix + "/post"
+        XCTAssertThrowsError(
+            try self.defaultClient.execute(request:
+                Request(url: url,
+                        body: .stream(length: 10) { streamWriter in
+                            let promise = self.defaultClient.eventLoopGroup.next().makePromise(of: Void.self)
+                            DispatchQueue(label: "content-length-test").async {
+                                streamWriter.write(.byteBuffer(ByteBuffer(string: "1"))).cascade(to: promise)
+                            }
+                            return promise.futureResult
+                        })).wait()) { error in
+            XCTAssertEqual(error as! HTTPClientError, HTTPClientError.bodyLengthMismatch)
+        }
+        // Quickly try another request and check that it works.
+        var response = try self.defaultClient.get(url: self.defaultHTTPBinURLPrefix + "get").wait()
+        let info = try response.body!.readJSONDecodable(RequestInfo.self, length: response.body!.readableBytes)
+        XCTAssertEqual(info!.connectionNumber, 1)
+        XCTAssertEqual(info!.requestNumber, 1)
+    }
+
+    // currently gets stuck because of #250 the server just never replies
+    func testContentLengthTooShortFails() throws {
+        let url = self.defaultHTTPBinURLPrefix + "/post"
+        let tooLong = "XBAD BAD BAD NOT HTTP/1.1\r\n\r\n"
+        XCTAssertThrowsError(
+            try self.defaultClient.execute(request:
+                Request(url: url,
+                        body: .stream(length: 1) { streamWriter in
+                            streamWriter.write(.byteBuffer(ByteBuffer(string: tooLong)))
+                                                    })).wait()) { error in
+            XCTAssertEqual(error as! HTTPClientError, HTTPClientError.bodyLengthMismatch)
+        }
+        // Quickly try another request and check that it works. If we by accident wrote some extra bytes into the
+        // stream (and reuse the connection) that could cause problems.
+        var response = try self.defaultClient.get(url: self.defaultHTTPBinURLPrefix + "get").wait()
+        let info = try response.body!.readJSONDecodable(RequestInfo.self, length: response.body!.readableBytes)
+        XCTAssertEqual(info!.connectionNumber, 1)
```

Comment:
this cannot be right, this will only be correct if you run this tests on its own. To get the right connection number, execute a random, uninteresting request beforehand and then add 1 to it.

Reply:
the way it was implemented before was actually not good, imho. If you use a global counter (as it was implemented before), then you cannot run test that do anything concurrently. I re-wrote that to have per-server/per-connection counters

Reply:
@artemredkin that's true, but connection counter is still global, no? We could attach the counters to the HTTPBin maybe?

Reply:
no, connection counter is not global anymore, every new instance of HTTPBin has its own connection counter and every new connection (ie handler) has its own request counter

Reply:
Ok, sounds good!